src/server/serve.rs

```diff
@@ -0,0 +1,132 @@
+use hyper::body::Incoming;
+use hyper::server::conn::http1;
+use hyper::service::service_fn;
+use hyper_util::rt::{TokioIo, TokioTimer};
+use std::convert::Infallible;
+use std::sync::Arc;
+use tokio::net::{TcpListener, TcpStream};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+use tokio::task;
+
+use super::accept;
+use super::backoff::Backoff;
+use crate::body::Pollable;
+use crate::router::Router;
+use crate::{Error, Request};
+
+/// The request type used by our `hyper` service. This is the type that we will
+/// use to create a `via::Request` that will be passed to the middleware stack.
+type HttpRequest = http::Request<Incoming>;
+
+/// The response type used by our `hyper` service. This is the type that we will
+/// unwrap from the `via::Response` returned from the middleware stack.
+type HttpResponse = http::Response<Pollable>;
+
+pub async fn serve<State>(
+    state: Arc<State>,
+    router: Arc<Router<State>>,
+    listener: TcpListener,
+    max_connections: usize,
+) -> Result<(), Error>
+where
+    State: Send + Sync + 'static,
+{
+    // Create a new backoff strategy with a base delay of 20 milliseconds and a
+    // maximum delay of 30 seconds.
+    let mut backoff = Backoff::new(20, 30);
+    // Create a vector to store the join handles of the spawned tasks. We'll
+    // periodically check if any of the tasks have finished and remove them.
+    let mut handles = Vec::new();
```

Comment:
Keeping a reference to join handles until the task has finished could be a useful metric for tracing in the near future.